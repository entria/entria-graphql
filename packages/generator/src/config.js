/* @flow */
import path from 'path';
import merge from 'lodash.merge';
import fs from 'fs';
import pkgDir from 'pkg-dir';

const rootPath = pkgDir.sync('.') || '.';

let cacheConfig = null;

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath: string) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === DIRECTORY_TYPE.SOURCE) {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
        [DIRECTORY_TYPE.SRC]: config.directories.source,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories.module}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

export const DIRECTORY_TYPE = {
  SOURCE: 'source',
  SRC: 'src',
  MODULE: 'module',
  CONNECTION: 'connection',
  LOADER: 'loader',
  MODEL: 'model',
  MUTATION: 'mutation',
  MUTATION_TEST: 'mutation_test',
  TYPE: 'type',
  TYPE_TEST: 'type_test',
};

type DirectoryType = $Values<typeof DIRECTORY_TYPE>;
type Directories = {
  source: string,
  module: string,
  connection: string,
  loader: string,
  model: string,
  mutation: string,
  mutation_test: string,
  type: string,
  type_test: string,
  interface: string,
};

type Files = {
  schema: string,
};

type Config = {
  directories: Directories,
  files: Files,
}
/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = (): Config => {
  // if (cacheConfig) return cacheConfig;

  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const config = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    merge(config, customConfig);

    cacheConfig = config;

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return config;
  } catch (err) {
    cacheConfig = config;
    // Return the default config if the custom doesn't exist
    return config;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory: DirectoryType) => getCreateGraphQLConfig().directories[directory];
