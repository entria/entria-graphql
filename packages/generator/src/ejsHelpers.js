/* @flow */
/**
 * Uppercase the first letter of a text
 * @param text {string}
 * @returns {string}
 */
export const uppercaseFirstLetter = (text: string) => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;
/**
 * Camel cases text
 * @param text {string} Text to be camel-cased
 * @returns {string} Camel-cased text
 */
export const camelCaseText = (text: string) =>
  text.replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
    if (+match === 0) {
      return '';
    }

    return index === 0 ? match.toLowerCase() : match.toUpperCase();
  });
