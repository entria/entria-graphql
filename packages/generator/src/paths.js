/* @flow */
import path from 'path';

export const getModulePath = (source: string, moduleName: string) => {
  return path.join(source, moduleName);
};

export const getTestPath = (source: string) => {
  return path.join(source, '__tests__');
};

export const getMutationPath = (source: string) => {
  return path.join(source, 'mutation');
};
