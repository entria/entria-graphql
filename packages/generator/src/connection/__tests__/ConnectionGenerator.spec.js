import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';
import fs from 'fs-extra';

import {
  getFileContent,
  getFixturePath,
} from '../../../test/helpers';

import { getConfigDir } from '../../config';
import { uppercaseFirstLetter } from '../../ejsHelpers';
import { getModulePath } from '../../paths';

const connectionGenerator = path.join(__dirname, '..');

it('generate a connection', async () => {
  const moduleName = 'example';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(connectionGenerator)
    .withArguments(name)
    .toPromise();

  const destinationDir = getConfigDir('connection');

  const modulePath = getModulePath(destinationDir, moduleName);
  const connectionFilepath = path.join(modulePath, `${name}Connection.js`);

  assert.file([
    connectionFilepath,
  ]);

  const files = {
    connection: getFileContent(path.join(folder, connectionFilepath)),
  };

  expect(files).toMatchSnapshot();
});

it('generate a connection with schema', async () => {
  const moduleName = 'post';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(connectionGenerator)
    .inTmpDir(dir =>
      fs.copySync(
        getFixturePath(name),
        path.join(dir, 'src/modules/post/Post.js'),
      ),
    )
    .withArguments('Post Post')
    .toPromise();

  const destinationDir = getConfigDir('connection');
  const modulePath = getModulePath(destinationDir, moduleName);
  const connectionFilepath = path.join(modulePath, `${name}Connection.js`);

  assert.file([
    connectionFilepath,
  ]);

  const files = {
    connection: getFileContent(path.join(folder, connectionFilepath)),
  };

  expect(files).toMatchSnapshot();
});
