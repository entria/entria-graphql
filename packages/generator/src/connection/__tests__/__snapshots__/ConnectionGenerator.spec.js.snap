// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate a connection 1`] = `
Object {
  "connection": "// @flow
import { connectionDefinitions } from '../../graphql/connection/CustomConnectionType';

import ExampleType from './ExampleType';

export default connectionDefinitions({
  name: 'Example',
  nodeType: ExampleType,
});
",
}
`;

exports[`generate a connection with schema 1`] = `
Object {
  "connection": "// @flow
import { connectionDefinitions } from '../../graphql/connection/CustomConnectionType';

import PostType from './PostType';

export default connectionDefinitions({
  name: 'Post',
  nodeType: PostType,
});
",
}
`;
